src/api/v4/layer/layer.js

```diff
@@ -20,65 +22,122 @@ var CartoDBLayer = require('../../../geo/map/cartodb-layer');
+  this._internalModel = undefined;
+
+  this.id = id;
   this._source = source;
   this._style = style;
   this._visible = true;
+  this._featureClickColumns = options.featureClickColumns;
+  this._featureOverColumns = options.featureOverColumns;
 }
 
 Layer.prototype.setStyle = function (style, opts) {
   opts = opts || {};
   this._style = style;
-  if (!this._internalModel) {
-    return;
-  }
   this._internalModel.set('cartocss', style.toCartoCSS(), { silent: true });
-  if (opts.reload === false) {
-    return Promise.resolve();
+  return this._reloadEngine();
+};
+
+Layer.prototype.setSource = function (source) {
+  this._source = source;
+  if (this._internalModel) {
+    this._internalModel.set('source', source, { silent: true });
+  }
+  return this._reloadEngine();
+};
+
+Layer.prototype.setFeatureClickColumns = function (columns) {
+  this._featureClickColumns = columns;
+  if (this._internalModel) {
+    this._internalModel.infowindow.update(getInteractivityFields(columns), { silent: true });
   }
-  return this._internalModel._engine.reload();
+
+  return this._reloadEngine();
+};
+
+Layer.prototype.setFeatureOverColumns = function (columns) {
+  this._featureOverColumns = columns;
+  if (this._internalModel) {
+    this._internalModel.tooltip.update(getInteractivityFields(columns), { silent: true });
+  }
+
+  return this._reloadEngine();
 };
 
 Layer.prototype.hide = function () {
   this._visible = false;
   if (this._internalModel) {
-    this._internalModel.hide();
+    this._internalModel.set('visible', false, { silent: true });
   }
+
+  return this._reloadEngine();
 };
 
 Layer.prototype.show = function () {
   this._visible = true;
   if (this._internalModel) {
-    this._internalModel.show();
+    this._internalModel.set('visible', true, { silent: true });
   }
-};
 
-Layer.prototype.setSource = function (source) {
-  this._source = source;
-  if (this._internalModel) {
-    this._internalModel.set('source', source);
-  }
+  return this._reloadEngine();
 };
 
 Layer.prototype.$setEngine = function (engine) {
+  if (this._engine) {
+    return;
+  }
+  this._engine = engine;
   this._source.$setEngine(engine);
-  this._internalModel = new CartoDBLayer({
-    id: this._id,
+  this._internalModel = this._createInternalModel(engine);
+};
+
+Layer.prototype._createInternalModel = function (engine) {
+  return new CartoDBLayer({
+    id: this.id,
     source: this._source.$getInternalModel(),
     cartocss: this._style.toCartoCSS(),
-    visible: this._visible
+    visible: this._visible,
+    infowindow: getInteractivityFields(this._featureClickColumns),
+    tooltip: getInteractivityFields(this._featureOverColumns)
   }, { engine: engine });
 };
 
 Layer.prototype.$getInternalModel = function () {
   return this._internalModel;
 };
 
+Layer.prototype._reloadEngine = function () {
+  if (this._engine) {
+    return this._engine.reload();
+  }
+
+  return Promise.resolve();
+};
+
+_.extend(Layer.prototype, Backbone.Events);
```

Comment:
IMO we shoud use composition over exending prototipes.

In each case this line should be moved next to the contructor